### DISCUSSION TITLE
Is it possible to have browser pooling?

### DISCUSSION THREAD

**Discussion by sgirib** [gotenberg/gotenberg#647] [Q&A] [answered]
From the previous discussions, I reckon that Gotenberg uses a new browser instance per request when I use the Chromium module.

Similar to Puppeteer, is there any possibility of using a browser pool instead of spawning a new browser? I can extend it for such a need, though I want to initially check its feasibility.

I couldn't quite clearly understand the history of this approach with Gotenberg. From the previous discussions, it seems that earlier versions of Gotenberg used to pool browser instances. Am I mistaken?

**Comment by gulien** [comment]
Hello @sgirib 

Instead of a browsers pool, why not use the scalability of your container orchestrator to mimic such feature? 

IMO I could implement it, but there will be a good chance of memory leaks and the code base will definitely be harder to debug.

**Reply by gulien** [reply]
No such its worth the time investment.

**Comment by AndreasPaulsson** [comment]
I am also wondering about this and the reason is performance, mainly by 

1. Caching in the browser of assets (images, javascript etc) (or is this done currently? I am not sure if caching is enabled in the Chromium instance)
2. Startup time of the browser.

My use case is generation of invoices where we will generate a few thousand invoices at a time by converting HTML to PDF. All the invoices assets will be identical, there are only minor layout and/or text changes that differs between most documents. My idea is that reusing the same Chromium instance for a number of requests (10, 100, 1000?) would increase throughput a lot.

An idea would be to enable pooling with two parameters:

- Instance max request age which is the max number of requests that an Chromium instance should serve. After this number of requests has been served by a Chromium instance, the Chromium instance should be recycled. Default value is 1 which is same behavior as current. 
- Pool size which is number of Chromium instances. Default value is 1 which is same behavior as current. Maybe this should be solved at container orchestration level instead?

And I agree that this would make the code base harder to maintain so I would love a discussion of is this is worth it or not.
If this would fit the project, I would volunteer to make a first version. I am seasoned developer (30 years of professional development, mainly C++ and dotnet) but is pretty new to golang though so I would need some guidance.

**Reply by gulien** [reply, marked as answer]
Hello @AndreasPaulsson,

Due to #642, I'm actually working on a version with a single Chromium instance (with a restart mechanism like LibreOffice). I do think the pooling should definitely be done using the container orchestration level.

**Reply by AndreasPaulsson** [reply]
Yes, the more I think about, I do agree about the pooling. It should be handled at the container orchestration level.

But having instance reuse (similar to as LibreOffice seems to be using) would probably be useful to improve performance. Restarting when unhealthy would be nice but maybe hard to detect (unhealthy could mean a lot of different things) so maybe a request limit or similar would be useful.

Thank you for your answers, I think that you are doing a great work with this project.

**Reply by sgirib** [reply]
@AndreasPaulsson  just a quick response:  my use case is exactly the same. right now I am doing Puppeteer -> launch a chromium browser at the start of the process -> open a page for every new invoice as html -> print to pdf -> close page.

So the browser is launched and initialised once. Also multiple containers running a browser each, kafka messages distributing invoices to these browsers, every message opens a page and closes it once it is done.

I think, this is a reasonable approach.
